pallets/verifiers/src/tests.rs

```diff
@@ -95,32 +137,164 @@ mod register_should {
+        def_vk.execute_with(|| {
+            assert!(FakeVerifierPallet::vks(REGISTERED_VK_HASH).is_some());
+            FakeVerifierPallet::unregister_vk(RuntimeOrigin::signed(USER_1), REGISTERED_VK_HASH)
+                .unwrap();
+            assert!(FakeVerifierPallet::vks(REGISTERED_VK_HASH).is_none());
+        })
+    }
 
-    /// Provide an environment with a registered vk (use used the default vk)
-    #[fixture]
-    fn def_vk(mut test_ext: sp_io::TestExternalities) -> sp_io::TestExternalities {
-        test_ext.execute_with(|| {
-            FakeVerifierPallet::register_vk(RuntimeOrigin::signed(1), Box::new(REGISTERED_VK))
+    #[rstest]
+    fn keep_previously_registered_vk_around_if_another_user_is_referencing_it(
+        mut def_vk: sp_io::TestExternalities,
+    ) {
+        def_vk.execute_with(|| {
+            FakeVerifierPallet::register_vk(RuntimeOrigin::signed(USER_2), Box::new(REGISTERED_VK))
                 .unwrap();
-            System::reset_events();
-        });
-        test_ext
+            FakeVerifierPallet::unregister_vk(RuntimeOrigin::signed(USER_1), REGISTERED_VK_HASH)
+                .unwrap();
+            assert!(FakeVerifierPallet::vks(REGISTERED_VK_HASH).is_some());
+        })
+    }
+
+    #[rstest]
+    fn release_deposit(mut def_vk: sp_io::TestExternalities) {
+        def_vk.execute_with(|| {
+            assert!(Balances::reserved_balance(USER_1) > 0);
+            FakeVerifierPallet::unregister_vk(RuntimeOrigin::signed(USER_1), REGISTERED_VK_HASH)
+                .unwrap();
+            assert_eq!(Balances::reserved_balance(USER_1), 0)
+        })
+    }
+
+    #[rstest]
+    fn emit_event(mut def_vk: sp_io::TestExternalities) {
+        def_vk.execute_with(|| {
+            assert_ok!(FakeVerifierPallet::unregister_vk(
+                RuntimeOrigin::signed(USER_1),
+                REGISTERED_VK_HASH
+            ));
+            mock::System::assert_last_event(
+                Event::VkUnregistered {
+                    hash: REGISTERED_VK_HASH,
+                }
+                .into(),
+            );
+        })
     }
 
+    mod fail {
+        use super::*;
+        use frame_support::assert_noop;
+
+        #[rstest]
+        fn on_root_origin(mut test_ext: sp_io::TestExternalities) {
+            test_ext.execute_with(|| {
+                assert_noop!(
+                    FakeVerifierPallet::unregister_vk(RuntimeOrigin::root(), REGISTERED_VK_HASH),
+                    DispatchError::BadOrigin
+                );
+            })
+        }
+
+        #[rstest]
+        fn on_wrong_origin(mut def_vk: sp_io::TestExternalities) {
+            def_vk.execute_with(|| {
+                assert_noop!(
+                    FakeVerifierPallet::unregister_vk(
+                        RuntimeOrigin::signed(USER_2),
+                        REGISTERED_VK_HASH
+                    ),
+                    RError::VerificationKeyNotFound
```

Comment:
Looking at the test I was thinking that maybe better to raise a different Error here ? Something like "UserNotAuthorized" or "BadOrigin" ?

Reply:
Just changing the returned error would break the test `unregister_should::fail::on_nonexisting_vk`. So, instead in https://github.com/zkVerify/zkVerify/pull/178/commits/45d90e143cbb44855c43da3213fd743b8ebda58e I added the logic for distinguishing the two possible failure reasons (the vk does not exist, or it exists but was not registered by the caller).
The error handling logic introduces another branch in the code, which is not a "subset" of the execution path covered by the current benchmarks. However, the only expensive operation there is a storage read. And the happy path (covered by the benchmark) already performs storage reads and writes, so it should be strictly more expensive. Therefore, if we stick to a worst-case scenario benchmarking, it shouldn't be necessary to add an additional benchmark for the error handling path.